### PR TITLE
#12544: support wide channels (> 256) in maxpool

### PR DESCRIPTION
**Pull request by mywoodstock**
### Ticket
#12544 

### Problem description
Maxpool had limitation on max tiles for reduction to 8 (DST reg max). 

### What's changed
Added support for wide channels.

### Checklist
- [ ] Post commit CI passes
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [ ] Device performance regression CI testing passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/pool/maxpool/device/kernels/compute/max_pool_multi_core.cpp [hunk 1]
```diff
@@ -95,33 +84,32 @@ void MAIN {
...
     constexpr uint32_t out_cb_id = tt::CB::c_out0;
 
-    // const uint32_t TILE_WIDTH = 32;
     constexpr bool is_partial_tile = in_c < 32;
     static_assert((!is_partial_tile || (in_c == 16)), "Partial tile must have c_dim 16");
     constexpr uint32_t num_faces_in_tile = is_partial_tile ? 1 : 2;
     constexpr uint32_t num_out_rows = 1;
 
-    tilizeA_B_reduce_init<true>(in_cb_id, in_scalar_cb_id, in_ntiles_hwc, out_cb_id, num_faces_in_tile, window_size_hw);
+    uint32_t in_ntiles_hwc_block = in_ntiles_hwc / in_nblocks_c;
```
Comment (pavlejosipovic) [inline ttnn/cpp/ttnn/operations/pool/maxpool/device/kernels/compute/max_pool_multi_core.cpp:106]: maybe add constexpr here

File: ttnn/cpp/ttnn/operations/pool/maxpool/device/kernels/compute/max_pool_multi_core.cpp [hunk 2]
```diff
@@ -4,29 +4,20 @@
...
 
-// #include "compute_kernel_api.h"
 #include "compute_kernel_api/tilize.h"
 #include "compute_kernel_api/reduce.h"
 #include "compute_kernel_api/pack_untilize.h"
-// #include "tools/profiler/kernel_profiler.hpp"
 
 #define DEBUG_PRINT 0
 
 #if DEBUG_PRINT == 1
```
Comment (pavlejosipovic) [inline ttnn/cpp/ttnn/operations/pool/maxpool/device/kernels/compute/max_pool_multi_core.cpp:13]: feels like these should be in some common place for kernel debug utils

File: ttnn/cpp/ttnn/operations/pool/maxpool/device/kernels/compute/max_pool_multi_core.cpp [hunk 3]
```diff
@@ -40,47 +31,45 @@
...
+                << "num_pages: " << cb_interface[cb_id].fifo_num_pages << ", "
+                << "rd_ptr: " << cb_interface[cb_id].fifo_rd_ptr << ", "
+                << "wr_ptr: " << cb_interface[cb_id].fifo_wr_ptr << ", "
+                << "wr_tile_ptr: " << cb_interface[cb_id].fifo_wr_tile_ptr << " }" << ENDL();
+    }
 #endif
 
-template<uint32_t in_ntiles_hw, uint32_t in_ntiles_c, uint32_t out_ntiles_c, uint32_t nblocks, bool is_partial_tile, uint32_t split_reader, uint32_t unpA_face_r_dim>
+template<uint32_t in_ntiles_hw, uint32_t in_ntiles_c, uint32_t out_ntiles_c, bool is_partial_tile, uint32_t split_reader, uint32_t unpA_face_r_dim, uint32_t in_nblocks_c>
 inline void reduce_h_fused(
```
Comment (pavlejosipovic) [inline ttnn/cpp/ttnn/operations/pool/maxpool/device/kernels/compute/max_pool_multi_core.cpp:47]: "fused" in reduce_h_fused means tilize/untilize on the fly right?

**Review by pavlejosipovic — APPROVED** [review]

File: tests/ttnn/unit_tests/operations/test_maxpool2d.py [hunk 4]
```diff
@@ -113,7 +113,7 @@ def run_max_pool(
     # interleaved_mem_config = ttnn.L1_MEMORY_CONFIG
     # output = ttnn.to_memory_config(output, interleaved_mem_config)
     output_host = output.cpu()
-    output_pytorch_padded = ttnn.to_torch(output_host)
+    output_pytorch_padded = torch.Tensor(ttnn.to_torch(output_host))
```
Comment (ayerofieiev-tt) [inline tests/ttnn/unit_tests/operations/test_maxpool2d.py:116]: why is this needed? ttnn.to_torch returns torch.Tensor

**Review by ayerofieiev-tt — APPROVED** [review]

File: tests/ttnn/unit_tests/operations/test_maxpool2d.py (hunk 4 above)
Comment (mywoodstock) [reply tests/ttnn/unit_tests/operations/test_maxpool2d.py:116]: It doesn't really return pure torch tensor -- needs ttnn to subsequently read/manipulate it. This was the recommended way to convert it into pure torch.

File: ttnn/cpp/ttnn/operations/pool/maxpool/device/kernels/compute/max_pool_multi_core.cpp (hunk 3 above)
Comment (mywoodstock) [reply ttnn/cpp/ttnn/operations/pool/maxpool/device/kernels/compute/max_pool_multi_core.cpp:47]: yes

File: tests/ttnn/unit_tests/operations/test_maxpool2d.py (hunk 4 above)
Comment (dmakoviichuk-tt) [reply tests/ttnn/unit_tests/operations/test_maxpool2d.py:116]: Yeah can confirm. If you don't do it torch function will crash with segmentation fault.

File: tests/ttnn/unit_tests/operations/test_maxpool2d.py [hunk 5]
```diff
@@ -187,6 +189,10 @@ def run_max_pool(
             # [64, 16, 528, 80],  ## oom
             # [128, 16, 528, 80], ## oom
             # [256, 16, 528, 80], ## oom
+            ## wide for vgg
+            [1, 256, 56, 56],
+            [1, 512, 28, 28],
+            [1, 512, 14, 14],
```
Comment (dmakoviichuk-tt) [inline tests/ttnn/unit_tests/operations/test_maxpool2d.py:195]: is it possible to test it on multiple batches too?
also what if W or H is not equal to each other.
Also the case wher W or H is equal to 1 is interesting too.

Comment (mywoodstock) [reply tests/ttnn/unit_tests/operations/test_maxpool2d.py:195]: These are specific to what are used in models -- the generic testing will be all in sweep tests.

Comment (dmakoviichuk-tt) [reply tests/ttnn/unit_tests/operations/test_maxpool2d.py:195]: But if it works on other cases why not demonstrate it here?

File: tests/ttnn/unit_tests/operations/test_maxpool2d.py (hunk 4 above)
Comment (ayerofieiev-tt) [reply tests/ttnn/unit_tests/operations/test_maxpool2d.py:116]: 🫠 okay, good to know
